libs/clients/directorate-of-immigration/src/lib/directorateOfImmigrationClient.service.ts

```diff
@@ -0,0 +1,608 @@
+    auth: Auth,
+  ): Promise<CountryOfResidence[]> {
+    try {
+      const res = await this.countryOfResidenceApiWithAuth(
+        auth,
+      ).apiCountryOfResidenceGetAllGet()
+
+      return res.map((item) => ({
+        countryId: item.countryId?.toString() || '',
+        countryName: item.countryName || '',
+      }))
+    } catch (e) {
+      const errorMsg = `Error when trying to get country of residence from UTL for applicant ${auth.nationalId}`
+      this.logger.error(errorMsg, e)
+      throw new Error(errorMsg)
+    }
+  }
+
+  async getOldStayAbroadList(auth: Auth): Promise<StayAbroad[]> {
+    try {
+      const res = await this.residenceAbroadApiWithAuth(
+        auth,
+      ).apiResidenceAbroadGetAllGet()
+
+      return res.map((item) => ({
+        countryId: item.countryId?.toString() || '',
+        countryName: item.countryName || '',
+        dateFrom: item.dateFrom,
+        dateTo: item.dateTo,
+        purposeOfStay: item.purposeOfStay,
+      }))
+    } catch (e) {
+      const errorMsg = `Error when trying to get residence abroad from UTL for applicant ${auth.nationalId}`
+      this.logger.error(errorMsg, e)
+      throw new Error(errorMsg)
+    }
+  }
+
+  async getOldPassportItem(auth: Auth): Promise<Passport | undefined> {
+    try {
+      const res = await this.travelDocumentApiWithAuth(
+        auth,
+      ).apiTravelDocumentGetAllGet()
+
+      // Select the most recent entry
+      const newestItem = res.sort(
+        (a, b) => (b.createdOn?.getTime() || 0) - (a.createdOn?.getTime() || 0),
+      )[0]
+
+      return (
+        newestItem && {
+          dateOfIssue: newestItem.dateOfIssue,
+          dateOfExpiry: newestItem.dateOfExpiry,
+          name: newestItem.name,
+          passportNo: newestItem.travelDocumentNo,
+          passportTypeId: newestItem.travelDocumentTypeId,
+          passportTypeName: newestItem.travelDocumentTypeName,
+          issuingCountryId: newestItem.issuingCountryId?.toString(),
+          issuingCountryName: newestItem.issuingCountryName,
+        }
+      )
+    } catch (e) {
+      const errorMsg = `Error when trying to get users travel document from UTL for applicant ${auth.nationalId}`
+      this.logger.error(errorMsg, e)
+      throw new Error(errorMsg)
+    }
+  }
+
+  async getCitizenshipResidenceConditionInfo(
+    auth: Auth,
+  ): Promise<ResidenceConditionInfo> {
+    try {
+      const res = await this.applicantResidenceConditionApiWithAuth(
+        auth,
+      ).apiApplicantResidenceConditionGetGet()
+
+      return {
+        hasValid: res.isAnyResConValid || false,
+        hasOnlyTypeMaritalStatus:
+          res.isOnlyMarriedOrCohabitationWithISCitizen || false,
+      }
+    } catch (e) {
+      const errorMsg = `Error when trying to get citizenship residence conditions from UTL for applicant ${auth.nationalId}`
+      this.logger.error(errorMsg, e)
+      throw new Error(errorMsg)
+    }
+  }
+
+  async submitApplicationForCitizenship(
```

Comment:
This submit function is really big. It contains a lot of individual data posting to different endpoints in the UTL api, which all can fail. 

Doesn't that leave the data in some wierd state? This looks like a lot of business logic that UTL should handle in their API and only exposing a single endpoint for submitting application for citizenship.